File: java-backend/src/main/java/com/seekho/api/controller/StudentController.java
package com.seekho.api.controller;

import com.seekho.api.entity.Enquiry;
import com.seekho.api.entity.Student;
import com.seekho.api.repository.EnquiryRepo;
import com.seekho.api.service.StudentService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.client.RestTemplate;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
@CrossOrigin(origins = "http://localhost:5173")
@RestController
@RequestMapping("/api/student")
public class StudentController {

    @Autowired
    private StudentService studentService;

    @GetMapping
    public ResponseEntity<List<Student>> getAllStudents() {
        return new ResponseEntity<>(studentService.getAllStudents(), HttpStatus.OK);
    }

    @GetMapping("/{studentId}")
    public ResponseEntity<Student> getStudentById(@PathVariable int studentId) {
        Optional<Student> student = studentService.getStudentById(studentId);
        return student.map(value -> new ResponseEntity<>(value, HttpStatus.OK))
                .orElseGet(() -> ResponseEntity.notFound().build());
    }

//    @PostMapping
//    public ResponseEntity<String> addStudent(@RequestBody Student student) {
//        studentService.addStudent(student);
//        return ResponseEntity.status(HttpStatus.CREATED).body("Student added successfully");
//    }
    @Autowired
    private RestTemplate restTemplate;

    @Autowired
    private EnquiryRepo enquiryRepository;

    @PostMapping
    public ResponseEntity<String> addStudent(@RequestBody Student student) {
        studentService.addStudent(student);  // saves student

        // Safely fetch Enquiry
        int enquiryId = student.getEnquiry().getEnquiryId();
        Enquiry enquiry = enquiryRepository.findById(enquiryId).orElse(null);

        if (enquiry != null && enquiry.getEnquirerEmailId() != null) {
            Map<String, String> emailRequest = new HashMap<>();
            emailRequest.put("to", enquiry.getEnquirerEmailId());
            emailRequest.put("studentName", enquiry.getEnquirerName());

            try {
                restTemplate.postForObject("http://localhost:8081/api/mail/send", emailRequest, String.class);
            } catch (Exception e) {
                System.out.println("Mail service failed: " + e.getMessage());
            }
        } else {
            System.out.println("Enquiry or email was null, skipping mail.");
        }

        return ResponseEntity.status(HttpStatus.CREATED).body("Student added and email sent (if possible)");
    }



    @PutMapping
    public ResponseEntity<String> updateStudent(@RequestBody Student student) {
        boolean isUpdated = studentService.updateStudent(student);
        if (isUpdated) {
            return ResponseEntity.ok("Student updated");
        } else {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body("Student not found");
        }
    }

    @DeleteMapping("/{studentId}")
    public ResponseEntity<String> deleteStudent(@PathVariable int studentId) {
        studentService.deleteStudent(studentId);
        return ResponseEntity.ok("Student deleted");
    }
}

File: dotnet-backend/Controllers/StudentController.cs
using dotnet_backend.Models;
using dotnet_backend.Repositories;
using dotnet_backend.Services;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;

namespace dotnet_backend.Controllers
{
    [ApiController]
    [Route("api/[controller]")]
    public class StudentController : ControllerBase
    {
        private readonly IStudentService _studentService;
        private readonly AppDbContext _context;
        private readonly HttpClient _httpClient;

        public StudentController(IStudentService studentService, AppDbContext context, IHttpClientFactory httpClientFactory)
        {
            _studentService = studentService;
            _context = context;
            _httpClient = httpClientFactory.CreateClient();
        }

        [HttpGet]
        public ActionResult<IEnumerable<Student>> GetAll()
        {
            return Ok(_studentService.GetAllStudents());
        }

        [HttpGet("{studentId:int}")]
        public ActionResult<Student> GetById(int studentId)
        {
            var student = _studentService.GetStudentById(studentId);
            if (student == null) return NotFound();
            return Ok(student);
        }

        [HttpPost]
        public async Task<ActionResult<string>> Add([FromBody] Student student)
        {
            _studentService.AddStudent(student);

            if (student.EnquiryId.HasValue)
            {
                var enquiry = await _context.Enquiries.FirstOrDefaultAsync(e => e.EnquiryId == student.EnquiryId.Value);

                if (enquiry != null && !string.IsNullOrEmpty(enquiry.EnquirerEmailId))
                {
                    var emailPayload = new
                    {
                        to = enquiry.EnquirerEmailId,
                        studentName = enquiry.EnquirerName
                    };

                    try
                    {
                        await _httpClient.PostAsJsonAsync("http://localhost:8081/api/mail/send", emailPayload);
                    }
                    catch (Exception ex)
                    {
                        Console.WriteLine($"Mail service failed: {ex.Message}");
                    }
                }
            }

            return Created("", "Student added and email sent (if possible)");
        }

        [HttpPut]
        public IActionResult Update([FromBody] Student student)
        {
            if (_studentService.UpdateStudent(student))
                return Ok("Student updated");
            return NotFound("Student not found");
        }

        [HttpDelete("{studentId:int}")]
        public IActionResult Delete(int studentId)
        {
            _studentService.DeleteStudent(studentId);
            return Ok("Student deleted");
        }
    }
}
